File: 2020/day_13/day_13.py
import math

with open('in') as f:
    earliest_departure = int(f.readline())
    buses = list(map(lambda b: (b[0], int(b[1])), [b for b in enumerate(f.readline().split(',')) if b[1] != 'x']))


def part_one():
    departure = None
    bus = None

    for b in buses:
        t = math.ceil(earliest_departure / b[1]) * b[1]
        if departure is None or t < departure:
            departure = t
            bus = b[1]

    return bus * (departure - earliest_departure)


def part_two():
    N = 1
    for _, b in buses:
        N *= b

    t = 0
    for i, b in buses:
        Ni = N // b
        Mi = pow(Ni, -1, b)
        t += -i * Mi * Ni

    return t % N


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_13/day_13.rs
use std::fs;

fn main() {
    let data = fs::read_to_string("in").unwrap();
    let data: Vec<&str> = data.lines().collect();

    let earliest_departure: u32 = data[0].parse().unwrap();
    let buses: Vec<(usize, u32)> = data[1]
        .split(',')
        .enumerate()
        .filter(|(_, b)| *b != "x")
        .map(|(i, b)| (i, b.parse().unwrap()))
        .collect();

    println!("Part One: {}", part_one(earliest_departure, &buses));
    println!("Part Two: {}", part_two(&buses));
}

fn part_one(earliest_departure: u32, buses: &Vec<(usize, u32)>) -> u32 {
    let (b, t) = buses
        .iter()
        .map(|(_, b)| {
            let t = (earliest_departure as f64 / *b as f64).ceil() as u32 * b;
            (b, t)
        })
        .min_by_key(|(_, t)| *t)
        .unwrap();

    b * (t - earliest_departure)
}

fn part_two(buses: &Vec<(usize, u32)>) -> u64 {
    let n = buses.iter().fold(1u64, |acc, (_, b)| acc * (*b as u64));

    let res = buses.iter().fold(0u64, |acc, (i, b)| {
        let index = (-(*i as i32)).rem_euclid(*b as i32) as u64;

        let ni = n / (*b as u64);
        let mi = inverse(ni, (*b).into());

        acc + index * mi * ni
    });

    res.rem_euclid(n)
}

fn inverse(x: u64, n: u64) -> u64 {
    let mut r1 = n;
    let mut r2 = x.rem_euclid(n);
    let mut s1: i64 = 0;
    let mut s2: i64 = 1;

    while r2 != 0 {
        let r3 = r1 % r2;
        
        let q = (r1 / r2) as i64;
        let s3 = s1 - q * s2;

        r1 = r2;
        r2 = r3;
        
        s1 = s2;
        s2 = s3;
    }

    s1.rem_euclid(n as i64) as u64
}
